libwasmvm/src/iterator.rs

```diff
@@ -97,4 +114,60 @@ impl GoIter {
         };
         (result, gas_info)
     }
+
+    pub fn next_key(&mut self) -> BackendResult<Option<Vec<u8>>> {
+        self.next_key_or_val(self.vtable.next_key, "next_key")
+    }
+
+    pub fn next_value(&mut self) -> BackendResult<Option<Vec<u8>>> {
+        self.next_key_or_val(self.vtable.next_value, "next_value")
+    }
+
+    #[inline(always)]
+    fn next_key_or_val(
+        &mut self,
+        next_fn: Option<
+            extern "C" fn(
+                iterator_t,
+                *mut gas_meter_t,
+                *mut u64,
+                *mut UnmanagedVector, // output
+                *mut UnmanagedVector, // error message output
+            ) -> i32,
+        >,
+        fn_name: &str,
+    ) -> BackendResult<Option<Vec<u8>>> {
+        let Some(next) = next_fn else {
+            let result = Err(BackendError::unknown(format!(
+                "iterator vtable function '{fn_name}' not set"
+            )));
+            return (result, GasInfo::free());
+        };
```

Comment:
If we move this unwrapping code to the caller we don't need the `Option<` and `fn_name`. But I'm happy either way.